adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/lwm2m/LwM2MResourceDirectory.java

```diff
@@ -0,0 +1,336 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Future<CoapContext> createCoapContextForPost(final CoapExchange coapExchange, final Span span) {
+        return getAuthenticatedDevice(coapExchange)
+                .map(device -> CoapContext.fromRequest(coapExchange, device, device, null, span));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Future<CoapContext> createCoapContextForDelete(final CoapExchange coapExchange, final Span span) {
+        return getAuthenticatedDevice(coapExchange)
+                .map(device -> CoapContext.fromRequest(coapExchange, device, device, null, span));
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Handles a request to create or update a device's registration.
+     */
+    @Override
+    protected Future<?> handlePost(final CoapContext ctx) {
+
+        final Request request = ctx.getExchange().advanced().getRequest();
+        LOG.debug("processing POST request: {}", request);
+
+        // we only support confirmable messages
+        if (!ctx.isConfirmable()) {
+            return Future.failedFuture(new ClientErrorException(
+                    HttpURLConnection.HTTP_BAD_REQUEST,
+                    "LwM2M does not support CoAP NON requests"));
+        }
+
+        if (!ctx.isDeviceAuthenticated()) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
+                    "server does not support noSec security mode"));
+        }
+
+        final List<String> uri = ctx.getExchange().getRequestOptions().getUriPath();
+        if (uri.size() > 2) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND));
+        }
+
+        final String lwm2mVersion = ctx.getQueryParameter(QUERY_PARAM_LWM2M_VERSION);
+        final String endpoint = ctx.getQueryParameter(QUERY_PARAM_ENDPOINT);
+        final Long lifetime = Optional.ofNullable(ctx.getQueryParameter(QUERY_PARAM_LIFETIME))
+                .map(lt -> {
+                    try {
+                        return Long.valueOf(lt);
+                    } catch (final NumberFormatException e) {
+                        return -1L;
+                    }
+                })
+                .orElse(null);
+        final BindingMode bindingMode = Optional.ofNullable(ctx.getQueryParameter(QUERY_PARAM_BINDING_MODE))
+                .map(BindingMode::valueOf)
+                .orElse(BindingMode.U);
+        final Link[] objectLinks = Link.parse(request.getPayload());
+
+        final Promise<ResponseCode> result = Promise.promise();
+        if (uri.size() == 1) {
+            handleRegister(ctx, lwm2mVersion, endpoint, bindingMode, lifetime, objectLinks).onComplete(result);
+        } else {
+            // URI path can have at most 2 segments as already verified above
+            handleUpdate(ctx, bindingMode, lifetime, objectLinks, uri.get(1)).onComplete(result);
```

Comment:
Hard to see, at least 

```java
if (uri.size() > 2)
```

doesn't cover 

```java
if (uri.size() == 0)
```

FMPOV, that should be

```java
if (uri.size() > 2 || uri.size() < 1)
```

Reply:
uri.size also cannot be 0, it contains at least `rd`, otherwise we wouldn't be handling the request in this Resource, would we?

Reply:
Yes, you're right.